github-ajax.html

```diff
@@ -4,125 +4,435 @@
       },
-      _onError: function(e) {
-        shared.headerCache[this.refurl] = e.detail.request.xhr.getResponseHeader('ETag');
-        // Cache responses to save the rate limit of the GitHub api
-        if (e.detail.request.status === 304) {
-          this._retrievePreviousRequestFromCache();
+
+      _handle304: function(o) {
+        if (this.lastRequest === o.request) {
+          if (o.response.status === 304) {
+            if (!o.cachedResponse) throw Error('Received a 304, but found no matching response in cache!');
+            return o.cachedResponse;
+          } else {
+            return this._checkResponseForErrors(o.request, o.response)
+              .then(this._cacheResponse(o.request).bind(this))
+          }
+        }
+      },
+
+      _checkResponseForErrors: function(request, response) {
+        if (response.status === 200) {
+          return Promise.resolve(response);
         } else {
-          this.fire('error', e.detail, {
-            bubbles: this.bubbles
+          return Promise.reject({
+            message: response.statusText,
+            errorCode: response.status,
+            response: response
           });
         }
       },
-      _retrievePreviousRequestFromCache: function() {
-        this._setLastResponse(shared.responseCache[this.refurl + 'page' + this.page]);
-        // Inject our last response
-        this.lastRequest._setResponse(this.lastResponse);
-        this.fire('response', Object.create(this.lastRequest), {
-          bubbles: this.bubbles
-        });
+
+      _cacheResponse: function(request) { return function(response) {
+        if (!this._shouldUseCache(response)) {
+          return Promise.resolve(response);
+        } else {
+          var responseClone = response.clone();
+          return caches.open(this._cacheName)
+            .then(cache => cache.put(request, response))
+            .then(() => Promise.resolve(responseClone));
+        }
+      }},
+
+      _decodeResponseBody: function(response) {
+        if (!response) return;
+        if (this.handleAs === 'json') {
+          return response.json();
+        } else if (this.handleAs === 'text') {
+          return response.text();
+        } else if (this.handleAs === 'arrayBuffer') {
+          return response.arrayBuffer();
+        }
+        throw new Error('Unrecognized `handleAs` property: ' + this.handleAs);
+      },
+
+      _constructRequest: function(headers) {
+        return new Request(this.requestUrl, {
+            method: this.method,
+            headers: headers,
+            body: this._encodeBody(headers)
+          });
+      },
+
+      _shouldUseCache: function(r) {
+        const nonCachedContent = [
+          'application/vnd.github.VERSION.diff',
+          'application/vnd.github.black-cat-preview+json',
+          'application/vnd.github.VERSION.raw'
+        ]
+
+        if (r instanceof Request) {
+          for (var i = 0, l = nonCachedContent.length; i < l; i++) {
```

Comment:
This version is faster since it does not have to access the `length` property in each iteration